core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -170,64 +264,61 @@ public void run() {
    */
   @Override
   public boolean hasListeners() {
-    return !registeredListeners.isEmpty();
-  }
-
-  private final Set<EventListenerWrapper> registeredListeners = new CopyOnWriteArraySet<EventListenerWrapper>();
-  private final ExecutorService orderedDelivery;
-  private final ExecutorService unorderedDelivery;
-
-  private static final class EventListenerWrapper {
-    final CacheEventListener<?, ?> listener;
-    final CacheEventListenerConfiguration config;
-    
-    EventListenerWrapper(CacheEventListener<?, ?> listener, final EventFiring firing, final EventOrdering ordering, 
-        final EnumSet<EventType> forEvents) {
-      this.listener = listener;
-      this.config = new CacheEventListenerConfiguration() {
-        
-        @Override
-        public Class<CacheEventListenerProvider> getServiceType() {
-          return CacheEventListenerProvider.class;
-        }
-        
-        @Override
-        public EventOrdering orderingMode() {
-          return ordering;
-        }
-        
-        @Override
-        public EventFiring firingMode() {
-          return firing;
-        }
-        
-        @Override
-        public EnumSet<EventType> fireOn() {
-          return forEvents;
+    return !syncListenersSet.isEmpty() || !aSyncListenersSet.isEmpty();
+  }
+
+  @Override
+  public List<CacheConfigurationChangeListener> getConfigurationChangeListeners() {
+    List<CacheConfigurationChangeListener> configurationChangeListenerList = new ArrayList<CacheConfigurationChangeListener>();
+    configurationChangeListenerList.add(new CacheConfigurationChangeListener() {
+      @Override
+      public void cacheConfigurationChange(final CacheConfigurationChangeEvent event) {
+        if (event.getProperty().equals(CacheConfigurationProperty.ADDLISTENER)) {
+          EventListenerWrapper newListener = (EventListenerWrapper)event.getNewValue();
+          CacheEventListenerConfiguration listenerConfiguration = newListener.config;
+          registerCacheEventListener(newListener.getListener(), listenerConfiguration.orderingMode(),
+              listenerConfiguration.firingMode(), listenerConfiguration.fireOn());
+        } else if (event.getProperty().equals(CacheConfigurationProperty.REMOVELISTENER)) {
+          CacheEventListener<K, V> oldListener = (CacheEventListener)event.getOldValue();
+          deregisterCacheEventListener(oldListener);
         }
-      };
-    }
-    
-    @SuppressWarnings("unchecked")
-    <K, V> CacheEventListener<K, V> getListener() {
-      return (CacheEventListener<K, V>) listener;
-    }
+      }
+    });
+    return configurationChangeListenerList;
+  }
 
-    @Override
-    public int hashCode() {
-      return listener.hashCode();
+  @Override
+  public void fireAllEvents() {
```

Comment:
Objective – Indicates that the events created by the concerned thread are eligible for firing. And should now be fired

What it internally does:
•	Goes through the threadLocal’s eventList, beginning with the first and call the dispatchEvent() method on the event
•	And then, remove this (first element) from the eventList
•	Continues until the eventList is empty i.e. all events created by the thread have been sent to the dispatch method

Why it should work:
Continuing with the premise that each thread fires the events it created throughout the operation. Each thread goes through the eventList it possesses, processing events and then removing them from the list thus going over every event it created. Finally clearing up the threadLocal.